keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,309 @@
+
+### Label Restrictions
+
+The Kubelet will be prevented from updating pod labels or creating mirror pods with labels, except
+for whitelisted keys:
+
+- Any labels starting with `unrestricted.node.kubernetes.io/` are allowed.
+
+The Kubelet does not currently label pods, nor are there official label keys that apply to
+pods. However, there are a few labels that are commonly applied to system addons & static pods:
+
+- `component` (used by [kubeadm][kubeadm-labels])
+- `tier` (used by [kubeadm][kubeadm-labels])
+- `k8s-app` (common on [addons][addons-k8s-app])
+
+The `k8s-app` label is used to match controllers for system components, and therefore should be
+explicitly disallowed.
+
+It is not clear how the `tier` and `component` labels are consumed, but I recommend uses be migrated
+to `unrestricted.node.kubernetes.io/component` and `unrestricted.node.kubernetes.io/tier`, rather
+than special casing the existing labels.
+
+[kubeadm-labels]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cmd/kubeadm/app/util/staticpod/utils.go#L60
+[addons-k8s-app]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cluster/addons/kube-proxy/kube-proxy-ds.yaml#L23
+
+### Annotation Restrictions
+
+The Kubelet will be prevented from updating pod annotations or creating mirror pods with
+annotations, except for whitelisted keys:
+
+1. Any annotations starting with `unrestricted.node.kubernetes.io/` are allowed.
+2. Annotations the Kubelet currently uses are allowed:
+- `ConfigMirrorAnnotationKey = "kubernetes.io/config.mirror"`
+- `ConfigHashAnnotationKey = "kubernetes.io/config.hash"`
+- `ConfigFirstSeenAnnotationKey = "kubernetes.io/config.seen"`
+- `ConfigSourceAnnotationKey = "kubernetes.io/config.source"`
+3. Well-known annotations that may be used on static pods are allowed:
+- `PodPresetOptOutAnnotationKey = "podpreset.admission.kubernetes.io/exclude"
+- `SeccompPodAnnotationKey = "seccomp.security.alpha.kubernetes.io/pod"`
+- `SeccompContainerAnnotationKeyPrefix = "container.seccomp.security.alpha.kubernetes.io/"` (prefix
+  match)
+- `ContainerAnnotationKeyPrefix = "container.apparmor.security.beta.kubernetes.io/"` (prefix match)
+- `PreferAvoidPodsAnnotationKey = "scheduler.alpha.kubernetes.io/preferAvoidPods"`
+- `BootstrapCheckpointAnnotationKey = "node.kubernetes.io/bootstrap-checkpoint"`
+
+Note that annotation restrictions are applying the least-privilege principal, but are not directly
+mitigating a known threat.
+
+### OwnerReferences
+
+OwnerReferences cannot be updated through the `pod/status` subresource, but they can be set on
+mirror pods. With the new restrictions, mirror pods are only allowed a single owner reference (or
+none), and it must refer to the node:
+
+```go
+ metav1.OwnerReference{
+  APIVersion: "v1"
+  Kind: "Node"
+  Name: node.Name
+  UID:  node.UID
+  Controller: nil // or false
+  BlockOwnerDeletion: nil // all values allowed
```

Comment:
do we care about all owner references, or just ones with `blockOwnerDeletion: true` and/or `controller: true`? what problems are caused by `ownerReferences` with `controller: false` and `blockOwnerDeletion: null | false`?

Reply:
(note that since a uid is required on ownerReferences, the actual uses for ownerReferences in static pod manifests seem extremely narrow, I just want to make sure we have clear specific threats that are being mitigated)

Reply:
I don't know enough about how OwnerReferences are used to confidently say non-true `controller` and `blockOwnerDeletion` are safe. Without a good use case, blocking all owner references seems like the safest and most future-proof option.

If you disagree with this logic, I'm open to to loosening these restrictions to only validate those fields.

Reply:
Thinking about this more, garbage collection would interact poorly with static pods... the GC controller would delete the pod, then the kubelet would recreate it with the same owner reference, which would make GC delete it, which would ...

So I think the proposed limitation is fine.